### PR TITLE
Add ROIPooling2D function

### PR DESCRIPTION
**Pull request by wkentaro**
- [x] CPU forward/backward
- [x] GPU forward/backward
- [x] Test for CPU/GPU

### CODE REVIEW

**Comment by delta2323** [comment]
Thank you for your PR! Could you add to the reference to the paper?

**Comment by wkentaro** [comment]
I have added

1. License declaration
1. Link to the fast-rcnn paper

File: chainer/functions/pooling/roi_pooling_2d.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+
+    def forward_cpu(self, inputs):
+        bottom_data, bottom_rois = inputs
+        n_rois, channels, height, width = bottom_data.shape
+        top_data = numpy.empty((n_rois, channels, self.outh, self.outw),
+                               dtype=numpy.float32)
+        self.argmax_data = numpy.empty_like(top_data).astype(numpy.int32)
+
+        for i_roi in six.moves.range(n_rois):
+            idx, xmin, ymin, xmax, ymax = bottom_rois[i_roi]
```
Comment (delta2323) [inline chainer/functions/pooling/roi_pooling_2d.py]: From this line, we must impose `x_type.shape[0] == roi_type.shape[0]` in `check_type_forward`

**Comment by delta2323** [comment]
In CPU, forward and backward propagation handle sample data in the minibatch one by one. Can we process them all at once?

File: chainer/functions/pooling/roi_pooling_2d.py (hunk 1 above)
Comment (wkentaro) [inline chainer/functions/pooling/roi_pooling_2d.py]: I updated the commit.

**Comment by wkentaro** [comment]
> In CPU, forward and backward propagation handle sample data in the minibatch one by one. Can we process them all at once (although its priority lower than GPU mode)?

Actually, I don't have idea how to achieve it, because ROIPooling2D need to slice the image array.

File: chainer/functions/pooling/roi_pooling_2d.py [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+
+            for outh in six.moves.range(self.outh):
+                sliceh, lenh = _roi_pooling_slice(
+                    outh, strideh, height, ymin)
+                if sliceh.stop <= sliceh.start:
+                    continue
+                for outw in six.moves.range(self.outw):
+                    slicew, lenw = _roi_pooling_slice(
+                        outw, stridew, width, xmin)
+                    if slicew.stop <= slicew.start:
```
Comment (unnonouno) [inline chainer/functions/pooling/roi_pooling_2d.py]: Do you need to set `argmax_data` in this situation?

File: chainer/functions/pooling/roi_pooling_2d.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+            '''
+            raw float32 top_diff, raw int32 argmax_data, int32 num_rois,
+            float32 spatial_scale, int32 channels, int32 height, int32 width,
+            int32 pooled_height, int32 pooled_width, raw float32 bottom_rois
+            ''',
+            'float32 bottom_diff',
+            '''
+            int w = i % width;
+            int h = (i / width) % height;
+            int c = (i / width / height) % channels;
```
Comment (unnonouno) [inline chainer/functions/pooling/roi_pooling_2d.py]: ``i / (width * height)`` is better.

File: chainer/functions/pooling/roi_pooling_2d.py [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+            int num = i / width / height / channels;
```
Comment (unnonouno) [inline chainer/functions/pooling/roi_pooling_2d.py]: `i / (width * height * channels)` is better.

File: chainer/functions/pooling/roi_pooling_2d.py [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+    This function acts similarly to :class:`~functions.MaxPooling2D`, but
+    it computes the maximum of input spatial patch for each channel
+    with the region of interest.
+
+    Args:
+        x (~chainer.Variable): Input variable.
+        rois (~chainer.Variable): Input roi variable.
+        outh (int): Height of output image after pooled.
+        outw (int): Width of output image after pooled.
+        spatial_scale (float): scale of the roi is resized.
```
Comment (unnonouno) [inline chainer/functions/pooling/roi_pooling_2d.py]: scale -> Scale

File: tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    @condition.retry(3)
+    def test_forward_cpu(self):
+        self.check_forward(self.x, self.rois)
+
+    @attr.gpu
+    @condition.retry(3)
+    def test_forward_gpu(self):
+        self.check_forward(cuda.to_gpu(self.x), cuda.to_gpu(self.rois))
+
+    def check_backward(self, x_data, roi_data, y_grad):
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py]: Can you use `gradient_check.check_backward` ?

File: tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+from chainer.testing import attr
+from chainer.testing import condition
+
+
+class TestROIPooling2D(unittest.TestCase):
+
+    def setUp(self):
+        N = 4
+        n_channels = 3
+        self.x = numpy.random.randn(N, n_channels, 12, 8).astype(numpy.float32)
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py]: Is this test stable? I mean, roi pooling contains `max` operator. It causes unstable behavior when values in an input variable contains similar values. Please check unit test of max-pooling.

**Comment by unnonouno** [comment]
It looks good except some comments.

File: tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+        self.assertEqual(self.gy.shape, y_data.shape)
+
+    @condition.retry(3)
+    def test_forward_cpu(self):
+        self.check_forward(self.x, self.rois)
+
+    @attr.gpu
+    @condition.retry(3)
+    def test_forward_gpu(self):
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py]: Can you check if result of GPU is equal to result of CPU?

File: chainer/functions/pooling/roi_pooling_2d.py (hunk 2 above)
Comment (wkentaro) [inline chainer/functions/pooling/roi_pooling_2d.py]: No, I think.

File: tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py (hunk 6 above)
Comment (wkentaro) [inline tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py]: actually it fails as below:
```
======================================================================
FAIL: test_backward_gpu (test_roi_pooling_2d.TestROIPooling2D)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/wkentaro/Documents/pfnet/chainer/chainer/testing/condition.py", line 71, in wrapper
    fail()
  File "/home/wkentaro/Documents/pfnet/chainer/chainer/testing/condition.py", line 51, in fail
    instance.fail(msg)
AssertionError:
Fail: 3, Success: 0

The first error message:
Traceback (most recent call last):
  File "/home/wkentaro/Documents/pfnet/chainer/chainer/testing/condition.py", line 57, in <lambda>
    lambda: f(*args, **kwargs),
  File "/home/wkentaro/Documents/pfnet/chainer/tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py", line 101, in test_backward_gpu
    cuda.to_gpu(self.gy))
  File "/home/wkentaro/Documents/pfnet/chainer/tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py", line 80, in check_backward
    gradient_check.check_backward(y, (x_data, roi_data), y_grad)
  File "/home/wkentaro/Documents/pfnet/chainer/chainer/gradient_check.py", line 224, in check_backward
    assert_allclose(gx, x.grad, atol=atol, rtol=rtol)
  File "/home/wkentaro/Documents/pfnet/chainer/chainer/gradient_check.py", line 76, in assert_allclose
    y = cuda.to_cpu(utils.force_array(y))
  File "/home/wkentaro/Documents/pfnet/chainer/chainer/cuda.py", line 223, in to_cpu
    '\nActual type: {0}.'.format(type(array)))
TypeError: The array sent to cpu must be numpy.ndarray or cupy.ndarray.
Actual type: <type 'NoneType'>.
```
with below diff:
```
diff --git a/tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py b/tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py
index 1f0d9d5..c53536f 100644
--- a/tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py
+++ b/tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py
@@ -73,19 +73,22 @@ class TestROIPooling2D(unittest.TestCase):
     def check_backward(self, x_data, roi_data, y_grad):
         x = chainer.Variable(x_data)
         rois = chainer.Variable(roi_data)
-        y = functions.roi_pooling_2d(x, rois, outh=self.outh, outw=self.outw,
-                                     spatial_scale=self.spatial_scale)
-        y.grad = y_grad
-        y.backward()
-
-        xs = (x.data, rois.data)
-
-        def f():
-            func = y.creator
-            return func.forward(xs)
-
-        gx, _ = gradient_check.numerical_grad(f, xs, (y.grad,))
-        gradient_check.assert_allclose(cuda.to_cpu(gx), cuda.to_cpu(x.grad))
+        # y = functions.roi_pooling_2d(x, rois, outh=self.outh, outw=self.outw,
+        #                              spatial_scale=self.spatial_scale)
+        y = functions.ROIPooling2D(outh=self.outh, outw=self.outw,
+                                   spatial_scale=self.spatial_scale)
+        gradient_check.check_backward(y, (x_data, roi_data), y_grad)
+        # y.grad = y_grad
+        # y.backward()
+        #
+        # xs = (x.data, rois.data)
+        #
+        # def f():
+        #     func = y.creator
+        #     return func.forward(xs)
+        #
+        # gx, _ = gradient_check.numerical_grad(f, xs, (y.grad,))
+        # gradient_check.assert_allclose(cuda.to_cpu(gx), cuda.to_cpu(x.grad))
 
     @condition.retry(3)
     def test_backward_cpu(self):

```

**Comment by wkentaro** [comment]
Thank you for the review. I have added some commits and comments. Could you please review again?

File: tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py (hunk 8 above)
Comment (wkentaro) [inline tests/chainer_tests/functions_tests/pooling_tests/test_roi_pooling_2d.py]: Yeah, I added.

File: chainer/functions/pooling/roi_pooling_2d.py [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+def roi_pooling_2d(x, rois, outh, outw, spatial_scale):
+    """Spatial Region of Interest (ROI) pooling function.
+
+    This function acts similarly to :class:`~functions.MaxPooling2D`, but
+    it computes the maximum of input spatial patch for each channel
+    with the region of interest.
+
+    Args:
+        x (~chainer.Variable): Input variable.
+        rois (~chainer.Variable): Input roi variable.
```
Comment (t-abe) [inline chainer/functions/pooling/roi_pooling_2d.py]: In my opinion, we should add explanation of rois' format.

Comment (wkentaro) [inline chainer/functions/pooling/roi_pooling_2d.py]: Thank you for the suggestion. I updated the docs.

**Comment by unnonouno** [comment]
Thanks a lot! I'll merge this PR in the next minor release.

**Pull request merged**